File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/search/SearchTrackAdapter.java
package mx.dev.franco.automusictagfixer.ui.search;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.List;

import mx.dev.franco.automusictagfixer.R;
import mx.dev.franco.automusictagfixer.covermanager.CoverLoader;
import mx.dev.franco.automusictagfixer.persistence.room.Track;
import mx.dev.franco.automusictagfixer.ui.AudioHolder;
import mx.dev.franco.automusictagfixer.ui.main.DiffCallback;
import mx.dev.franco.automusictagfixer.utilities.ServiceUtils;

public class SearchTrackAdapter extends RecyclerView.Adapter<FoundItemHolder> implements
        Observer<List<Track>> {
    private static final String TAG = SearchTrackAdapter.class.getName();
    private ServiceUtils serviceUtils;
    private FoundItemHolder.ClickListener mListener;
    private AsyncListDiffer<Track> asyncListDiffer = new AsyncListDiffer<>(this, new DiffCallback());



    public SearchTrackAdapter(FoundItemHolder.ClickListener listener){
        mListener = listener;
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        serviceUtils = ServiceUtils.getInstance(recyclerView.getContext());
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        if (mListener instanceof AsyncListDiffer.ListListener)
            asyncListDiffer.removeListListener((AsyncListDiffer.ListListener<Track>) mListener);

        serviceUtils = null;
        mListener = null;
        CoverLoader.cancelAll();
    }

    @Override
    public FoundItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).
                inflate(R.layout.result_search_item, parent, false);
        return new FoundItemHolder(itemView, mListener);
    }

    @Override
    public void onBindViewHolder(final FoundItemHolder holder, final int position) {
        Track track = asyncListDiffer.getCurrentList().get(position);
        enqueue(holder, track);
        holder.trackName.setText(track.getTitle());
        holder.artistName.setText(track.getArtist());
        holder.albumName.setText(track.getAlbum());

    }

    @Override
    public void onBindViewHolder(@NonNull final FoundItemHolder holder, int position, List<Object> payloads) {
        if(payloads.isEmpty()) {
            super.onBindViewHolder(holder,position,payloads);
        }
        else {
            Track track = asyncListDiffer.getCurrentList().get(position);
            Bundle o = (Bundle) payloads.get(0);
            for (String key : o.keySet()) {
                if (key.equals("title")) {
                    holder.trackName.setText(track.getTitle());
                }
                if (key.equals("artist")) {
                    holder.artistName.setText(track.getArtist());
                }
                if (key.equals("album")) {
                    holder.albumName.setText(track.getAlbum());
                }

                if (key.equals("should_reload_cover")){
                    enqueue(holder, track);
                }
            }
        }
    }

    private void enqueue(AudioHolder holder, Track track) {
        CoverLoader.startFetchingCover(holder, track.getPath(), track.getMediaStoreId()+"");
    }

    @Override
    public void onViewRecycled(@NonNull FoundItemHolder holder) {
        if(holder.itemView.getContext() != null)
            Glide.with(holder.itemView.getContext()).clear(holder.cover);
    }

    /**
     * Get size of data source
     * @return zero if data source is null, otherwise size of data source
     */
    @Override
    public int getItemCount() {
        if(asyncListDiffer != null)
            return asyncListDiffer.getCurrentList().size();
        return 0;
    }

    /**
     * Indicates whether each item in the data set
     * can be represented with a unique identifier of type Long.
     * @param hasStableIds true
     */
    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(true);
    }

    @Override
    public void onChanged(@Nullable List<Track> tracks) {
        asyncListDiffer.submitList(tracks);
    }
}

File: app/src/main/java/mx/dev/franco/automusictagfixer/ui/search/SearchTrackAdapter.kt
package mx.dev.franco.automusictagfixer.ui.search

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.AsyncListDiffer.ListListener
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import mx.dev.franco.automusictagfixer.R
import mx.dev.franco.automusictagfixer.covermanager.CoverLoader
import mx.dev.franco.automusictagfixer.persistence.room.Track
import mx.dev.franco.automusictagfixer.ui.AudioHolder
import mx.dev.franco.automusictagfixer.ui.main.DiffCallback
import mx.dev.franco.automusictagfixer.utilities.ServiceUtils

class SearchTrackAdapter(private var mListener: FoundItemHolder.ClickListener?) :
    RecyclerView.Adapter<FoundItemHolder>(), Observer<List<Track>?> {
    private var serviceUtils: ServiceUtils? = null
    private val asyncListDiffer: AsyncListDiffer<Track>? = AsyncListDiffer(this, DiffCallback())
    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        serviceUtils = ServiceUtils.getInstance(recyclerView.context)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        if (mListener is ListListener<*>) asyncListDiffer!!.removeListListener((mListener as ListListener<Track>?)!!)
        serviceUtils = null
        mListener = null
        CoverLoader.cancelAll()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FoundItemHolder {
        val itemView =
            LayoutInflater.from(parent.context).inflate(R.layout.result_search_item, parent, false)
        return FoundItemHolder(itemView, mListener)
    }

    override fun onBindViewHolder(holder: FoundItemHolder, position: Int) {
        val track = asyncListDiffer!!.currentList[position]
        enqueue(holder, track)
        holder.trackName.text = track.title
        holder.artistName.text = track.artist
        holder.albumName.text = track.album
    }

    override fun onBindViewHolder(holder: FoundItemHolder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads)
        } else {
            val track = asyncListDiffer!!.currentList[position]
            val o = payloads[0] as Bundle
            for (key in o.keySet()) {
                if (key == "title") {
                    holder.trackName.text = track.title
                }
                if (key == "artist") {
                    holder.artistName.text = track.artist
                }
                if (key == "album") {
                    holder.albumName.text = track.album
                }
                if (key == "should_reload_cover") {
                    enqueue(holder, track)
                }
            }
        }
    }

    private fun enqueue(holder: AudioHolder, track: Track) {
        CoverLoader.startFetchingCover(holder, track.path, track.mediaStoreId.toString() + "")
    }

    override fun onViewRecycled(holder: FoundItemHolder) {
        if (holder.itemView.context != null) Glide.with(holder.itemView.context).clear(holder.cover)
    }

    /**
     * Get size of data source
     * @return zero if data source is null, otherwise size of data source
     */
    override fun getItemCount(): Int {
        return asyncListDiffer?.currentList?.size ?: 0
    }

    /**
     * Indicates whether each item in the data set
     * can be represented with a unique identifier of type Long.
     * @param hasStableIds true
     */
    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(true)
    }

    override fun onChanged(tracks: List<Track>?) {
        asyncListDiffer!!.submitList(tracks)
    }

    companion object {
        private val TAG = SearchTrackAdapter::class.java.name
    }
}